### ISSUE TITLE
Can't build on alpine

### ISSUE THREAD

**Issue by michele** [OpenSMTPD/OpenSMTPD#983]
Hello,
I'm trying to build opensmtp on alpine linux: I tried both  HEAD and the portable tar.gz from the website, but when running make I get this:

```
gcc -DHAVE_CONFIG_H -I. -I../..  -I../../smtpd -I../../openbsd-compat -I../../openbsd-compat/err_h -I../../openbsd-compat/paths_h -I. -I/usr/include  -DSMTPD_CONFDIR=\"/usr/local/etc\" -DPATH_CHROOT=\"/var/empty\" -DPATH_SMTPCTL=\"/usr/local/sbin/smtpctl\" -DPATH_MAILLOCAL=\"/usr/local/libexec/opensmtpd/mail.local\" -DPATH_LIBEXEC=\"/usr/local/l
ibexec/opensmtpd\" -DHAVE_CONFIG_H -DIO_TLS -DCA_FILE=\"/etc/ssl/cert.pem\" -g -O2  -fPIC -DPIC -Wall -Wpointer-arith -Wuninitialized -Wsign-compare -Wformat-security -Wsizeof-pointer-memaccess -Wno-pointer-sign -Wno-unused-result -fno-strict-aliasing -fno-builtin-memset -D_BSD_SOURCE -D_DEFAULT_SOURCE  -D_GNU_SOURCE -DNEED_EVENT_ASR_RUN -c -o .
./../smtpd/smtpd-ca.o `test -f '../../smtpd/ca.c' || echo './'`../../smtpd/ca.c
../../smtpd/ca.c: In function 'rsae_pub_enc':
../../smtpd/ca.c:438:10: warning: implicit declaration of function 'RSA_meth_get_pub_enc' [-Wimplicit-function-declaration]
  return (RSA_meth_get_pub_enc(rsa_default)(flen, from, to, rsa, padding));
          ^~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:438:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'rsae_pub_dec':
../../smtpd/ca.c:446:10: warning: implicit declaration of function 'RSA_meth_get_pub_dec' [-Wimplicit-function-declaration]
  return (RSA_meth_get_pub_dec(rsa_default)(flen, from, to, rsa, padding));
          ^~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:446:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'rsae_priv_enc':
../../smtpd/ca.c:457:10: warning: implicit declaration of function 'RSA_meth_get_priv_enc' [-Wimplicit-function-declaration]
  return (RSA_meth_get_priv_enc(rsa_default)(flen, from, to, rsa, padding));
          ^~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:457:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'rsae_priv_dec':
../../smtpd/ca.c:469:10: warning: implicit declaration of function 'RSA_meth_get_priv_dec' [-Wimplicit-function-declaration]
  return (RSA_meth_get_priv_dec(rsa_default)(flen, from, to, rsa, padding));
          ^~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:469:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'rsae_mod_exp':
../../smtpd/ca.c:476:10: warning: implicit declaration of function 'RSA_meth_get_mod_exp' [-Wimplicit-function-declaration]
  return (RSA_meth_get_mod_exp(rsa_default)(r0, I, rsa, ctx));
          ^~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:476:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'rsae_bn_mod_exp':
../../smtpd/ca.c:484:10: warning: implicit declaration of function 'RSA_meth_get_bn_mod_exp' [-Wimplicit-function-declaration]
  return (RSA_meth_get_bn_mod_exp(rsa_default)(r, a, p, m, ctx, m_ctx));
          ^~~~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:484:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'rsae_init':
../../smtpd/ca.c:491:6: warning: implicit declaration of function 'RSA_meth_get_init' [-Wimplicit-function-declaration]
  if (RSA_meth_get_init(rsa_default) == NULL)
      ^~~~~~~~~~~~~~~~~
../../smtpd/ca.c:491:37: warning: comparison between pointer and integer
  if (RSA_meth_get_init(rsa_default) == NULL)
                                     ^~
../../smtpd/ca.c:493:10: error: called object is not a function or function pointer
  return (RSA_meth_get_init(rsa_default)(rsa));
          ^~~~~~~~~~~~~~~~~
../../smtpd/ca.c: In function 'rsae_finish':
../../smtpd/ca.c:500:6: warning: implicit declaration of function 'RSA_meth_get_finish' [-Wimplicit-function-declaration]
  if (RSA_meth_get_finish(rsa_default) == NULL)
      ^~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:500:39: warning: comparison between pointer and integer
  if (RSA_meth_get_finish(rsa_default) == NULL)
                                       ^~
../../smtpd/ca.c:502:10: error: called object is not a function or function pointer
  return (RSA_meth_get_finish(rsa_default)(rsa));
          ^~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c: In function 'rsae_keygen':
../../smtpd/ca.c:509:10: warning: implicit declaration of function 'RSA_meth_get_keygen' [-Wimplicit-function-declaration]
  return (RSA_meth_get_keygen(rsa_default)(rsa, bits, e, cb));
          ^~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:509:10: error: called object is not a function or function pointer
../../smtpd/ca.c: In function 'ECDSA_METHOD_new_temporary':
../../smtpd/ca.c:530:33: error: dereferencing pointer to incomplete type 'ECDSA_METHOD {aka struct ecdsa_method}'
  if ((ecdsa = calloc(1, sizeof (*ecdsa))) == NULL)
                                 ^~~~~~
../../smtpd/ca.c: In function 'ecdsae_do_sign':
../../smtpd/ca.c:623:23: error: dereferencing pointer to incomplete type 'ECDSA_METHOD {aka const struct ecdsa_method}'
  return (ecdsa_default->ecdsa_do_sign(dgst, dgst_len, inv, rp, eckey));
                       ^~
../../smtpd/ca.c: In function 'rsa_engine_init':
../../smtpd/ca.c:649:21: warning: implicit declaration of function 'RSA_meth_new' [-Wimplicit-function-declaration]
  if ((rsae_method = RSA_meth_new("RSA privsep engine", 0)) == NULL) {
                     ^~~~~~~~~~~~
../../smtpd/ca.c:649:19: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  if ((rsae_method = RSA_meth_new("RSA privsep engine", 0)) == NULL) {
                   ^
../../smtpd/ca.c:654:2: warning: implicit declaration of function 'RSA_meth_set_pub_enc' [-Wimplicit-function-declaration]
  RSA_meth_set_pub_enc(rsae_method, rsae_pub_enc);
  ^~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:655:2: warning: implicit declaration of function 'RSA_meth_set_pub_dec' [-Wimplicit-function-declaration]
  RSA_meth_set_pub_dec(rsae_method, rsae_pub_dec);
  ^~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:655:2: warning: implicit declaration of function 'RSA_meth_set_pub_dec' [-Wimplicit-function-declaration]                                                                                                                                                                                                                                 
  RSA_meth_set_pub_dec(rsae_method, rsae_pub_dec);                                                                                                                                                                                                                                                                                                         
  ^~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                                                                                                                                                     
../../smtpd/ca.c:656:2: warning: implicit declaration of function 'RSA_meth_set_priv_enc' [-Wimplicit-function-declaration]                                                                                                                                                                                                                                
  RSA_meth_set_priv_enc(rsae_method, rsae_priv_enc);                                                                                                                                                                                                                                                                                                       
  ^~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                                                                                                                                                    
../../smtpd/ca.c:657:2: warning: implicit declaration of function 'RSA_meth_set_priv_dec' [-Wimplicit-function-declaration]                                                                                                                                                                                                                                
  RSA_meth_set_priv_dec(rsae_method, rsae_priv_dec);                                                                                                                                                                                                                                                                                                       
  ^~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:658:2: warning: implicit declaration of function 'RSA_meth_set_mod_exp' [-Wimplicit-function-declaration]
  RSA_meth_set_mod_exp(rsae_method, rsae_mod_exp);
  ^~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:659:2: warning: implicit declaration of function 'RSA_meth_set_bn_mod_exp' [-Wimplicit-function-declaration]
  RSA_meth_set_bn_mod_exp(rsae_method, rsae_bn_mod_exp);
  ^~~~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:660:2: warning: implicit declaration of function 'RSA_meth_set_init' [-Wimplicit-function-declaration]
  RSA_meth_set_init(rsae_method, rsae_init);
  ^~~~~~~~~~~~~~~~~
../../smtpd/ca.c:661:2: warning: implicit declaration of function 'RSA_meth_set_finish' [-Wimplicit-function-declaration]
  RSA_meth_set_finish(rsae_method, rsae_finish);
  ^~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:662:2: warning: implicit declaration of function 'RSA_meth_set_keygen' [-Wimplicit-function-declaration]
  RSA_meth_set_keygen(rsae_method, rsae_keygen);
  ^~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:669:27: warning: implicit declaration of function 'RSA_meth_get0_name' [-Wimplicit-function-declaration]
   if (!ENGINE_set_name(e, RSA_meth_get0_name(rsae_method))) {
                           ^~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:669:27: warning: passing argument 2 of 'ENGINE_set_name' makes pointer from integer without a cast [-Wint-conversion]
In file included from ../../smtpd/ca.c:42:0:
/usr/include/openssl/engine.h:554:5: note: expected 'const char *' but argument is of type 'int'
 int ENGINE_set_name(ENGINE *e, const char *name);
     ^~~~~~~~~~~~~~~
../../smtpd/ca.c:687:40: warning: comparison between pointer and integer
  if (RSA_meth_get_mod_exp(rsa_default) == NULL)
                                        ^~
../../smtpd/ca.c:689:43: warning: comparison between pointer and integer
  if (RSA_meth_get_bn_mod_exp(rsa_default) == NULL)
                                           ^~
../../smtpd/ca.c:691:39: warning: comparison between pointer and integer
  if (RSA_meth_get_keygen(rsa_default) == NULL)
                                       ^~
../../smtpd/ca.c:693:2: warning: implicit declaration of function 'RSA_meth_set_flags' [-Wimplicit-function-declaration]
  RSA_meth_set_flags(rsae_method,
  ^~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:694:3: warning: implicit declaration of function 'RSA_meth_get_flags' [-Wimplicit-function-declaration]
   RSA_meth_get_flags(rsa_default) | RSA_METHOD_FLAG_NO_CHECK);
   ^~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:695:2: warning: implicit declaration of function 'RSA_meth_set0_app_data' [-Wimplicit-function-declaration]
  RSA_meth_set0_app_data(rsae_method,
  ^~~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c:696:3: warning: implicit declaration of function 'RSA_meth_get0_app_data' [-Wimplicit-function-declaration]
   RSA_meth_get0_app_data(rsa_default));
   ^~~~~~~~~~~~~~~~~~~~~~
../../smtpd/ca.c: In function 'rsae_keygen':
../../smtpd/ca.c:510:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_finish':
../../smtpd/ca.c:503:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_init':
../../smtpd/ca.c:494:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_bn_mod_exp':
../../smtpd/ca.c:485:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_mod_exp':
../../smtpd/ca.c:477:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_priv_dec':
../../smtpd/ca.c:470:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_priv_enc':
../../smtpd/ca.c:458:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_pub_dec':
../../smtpd/ca.c:447:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
../../smtpd/ca.c: In function 'rsae_pub_enc':
../../smtpd/ca.c:439:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make[3]: Leaving directory '/opensmtpd-6.6.0p1/mk/smtpd'
make[3]: *** [Makefile:767: ../../smtpd/smtpd-ca.o] Error 1
make[2]: *** [Makefile:372: all-recursive] Error 1
make[2]: Leaving directory '/opensmtpd-6.6.0p1/mk'
make[1]: *** [Makefile:419: all-recursive] Error 1
make[1]: Leaving directory '/opensmtpd-6.6.0p1'
make: *** [Makefile:351: all] Error 2
```

Tried fiddling with deps, but nothing seems to help.

**Comment by ngortheone** [comment]
@michele it looks like you are trying to build with LibreSSL.

Unfortunately LibreSSL on alpine is not supported and outdated. (Minimum required version is >=3, while alpine has 2.9.2)

replace libress with openssl and you should be fine.
Alpine Linux build is incorporated in our CI, you can see how we build it here:
https://github.com/OpenSMTPD/OpenSMTPD/blob/portable/ci/docker/Dockerfile.alpine

**Comment by poolpOrg** [comment]
Details as to why are available in a blog post in case you're interested: https://poolp.org/posts/2019-07-27/july-2019-report-tons-of-smtpd-work-mostly/

The changes I committed to LibreSSL are only available after 3.0.1

**Comment by michele** [comment]
@ngortheone @poolpOrg I was already targeting openssl as I had read about issues with libressl. The issue (and main difference with your Dockerfile), was that I wasn't selecting alpine 3.10, but building without a tag (which I shouldn't have done in the first place). Thanks a lot for your help!
